api/routes/image_pack.go

```diff
@@ -0,0 +1,197 @@
+//
+//   Copyright © 2020 Uncharted Software Inc.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+package routes
+
+import (
+	"bytes"
+	"encoding/json"
+	"image"
+	"image/draw"
+	"io/ioutil"
+	"net/http"
+	"path"
+
+	"github.com/nfnt/resize"
+	"github.com/pkg/errors"
+	"github.com/uncharted-distil/distil-compute/metadata"
+	"github.com/uncharted-distil/distil-compute/model"
+	"github.com/uncharted-distil/distil-compute/primitive/compute"
+	"github.com/uncharted-distil/distil/api/env"
+	api "github.com/uncharted-distil/distil/api/model"
+	"github.com/uncharted-distil/distil/api/util"
+)
+
+// ImagePackRequest is the expected request format for the route (Band can be an empty string)
+type ImagePackRequest struct {
+	Dataset  string   `json:"dataset"`
+	ImageIDs []string `json:"imageIds"`
+	Band     string   `json:"band,omitempty"`
+}
+
+// ImagePackResult is the expected post result for MultiBandImagePackHandler
+type ImagePackResult struct {
+	ImagesBuffer [][]byte `json:"images"`
+	ImageIDs     []string `json:"imageIds"`
+}
+type chanStruct struct {
+	data [][]byte
+	IDs  []string
+}
+
+func postParamsToImagePackRequest(r *http.Request) (*ImagePackRequest, error) {
+	body, err := ioutil.ReadAll(r.Body)
+	defer r.Body.Close()
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to parse POST request")
+	}
+	result := &ImagePackRequest{}
+	err = json.Unmarshal(body, result)
+	return result, err
+}
+
+// MultiBandImagePackHandler fetches individual band images and combines them into a single RGB image using the supplied mapping.
+func MultiBandImagePackHandler(ctor api.MetadataStorageCtor, dataCtor api.DataStorageCtor, config env.Config) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		params, err := postParamsToImagePackRequest(r)
+		if err != nil {
+			handleError(w, err)
+			return
+		}
+		// default to getImages
+		funcPointer := getImages
+		if params.Band != "" {
+			// if band is not empty then get multiBandImages
+			funcPointer = getMultiBandImages
+		}
+		// channel for threads to communicate
+		result := make(chan chanStruct)
+		// ImageThreadPool is an environment variable defaults to 2 (works great with 6)
+		for i := 0; i < config.ImageThreadPool; i++ {
+			go funcPointer(params, i, config.ImageThreadPool, result, ctor, dataCtor)
```

Comment:
It may be better to have the function accept an input channel. The go routine would then listen on that channel for messages (image ids?) and would process them, writing the result to the output channel. The main process would start the go routines, write the image ids (could be done in batches) to the input channel and listen to the output channel for results.

There's also the error handling dimension that needs minor changes. They appear to be generally ignored in the two functions. I am not too sure how we want the endpoint to behave if errors occur (fail the whole batch? return only successes? return all ids with empty results for errors?), but at the very least errors should be counted and a single log entry written if any error occurs (logging the amount of errors and the last error encountered or something like that).